### PR TITLE
Add validations and other form config | Closes #2427

### LINKED ISSUE #2427: Apply Remaining Form Config Options to New Edit Entity Forms
There are many additional input options that have not been applied to the inputs in the new edit date form nor the new edit ticket form. These include input options such as:

- required

- validations: required, minLength, valid dates, start dates before end dates, etc

- min, max, abd step values for number inputs

- help text

see:

- `ZZZ\editor\events\dates-and-times\editor-date\edit-form\use-date-entity-input-config.js`

- `ZZZ\editor\events\tickets\editor-ticket\edit-form\use-ticket-form-input-config.js`

for a list of all of the date and ticket entity properties and the various input configuration options applied to inputs in EDTR v1

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR:
- Adds validations to entity edit forms using `yup`
- Adds the remaining configuration options to the forms.
- Closes #2427

### CODE REVIEW

File: assets/src/application/ui/forms/espressoForm/utils.ts [hunk 1]
```diff
@@ -79,3 +80,21 @@ export const evalFieldConditions = (conditions: FieldConditions, formData: AnyOb
...
+ * @param values
+ */
+export const yupToFinalFormErrors = async <T>(validationSchema: ObjectSchema, values: T) => {
+	try {
+		await validationSchema.validate(values, { abortEarly: false });
+	} catch (error) {
+		return (error as ValidationError).inner.reduce(
+			(formError, innerError) => setIn(formError, innerError.path, innerError.message),
+			{}
+		);
```
Comment (nicolad) [inline assets/src/application/ui/forms/espressoForm/utils.ts]: can this be written a bit more explicit? not clear what is happening at first glance

File: assets/src/application/ui/forms/formModal/FormModalForm.tsx [hunk 2]
```diff
@@ -20,9 +20,18 @@ const modalStyle: CSSProperties = {
...
+	hasValidationErrors,
+	hasSubmitErrors,
+	pristine,
+	onClose,
+	children,
+	title,
+}) => {
 	const submitButton: ButtonProps = {
-		disabled: submitting || pristine,
+		disabled: submitting || pristine || hasValidationErrors || hasSubmitErrors,
```
Comment (nicolad) [inline assets/src/application/ui/forms/formModal/FormModalForm.tsx]: pulling this out might be cleaner

File: assets/src/domain/eventEditor/ui/datetimes/dateForm/useDateFormConfig.ts [hunk 3]
```diff
@@ -26,8 +19,16 @@ const FIELD_NAMES: Array<keyof Datetime> = ['name', 'description', 'capacity', '
...
+	let startDate = PLUS_ONE_MONTH;
+	let endDate = PLUS_TWO_MONTHS;
+
+	if (start) {
+		// "start" and "end" may be from Optimistic response as Date
+		startDate = (start as any) instanceof Date ? (start as any) : parseISO(start);
+	}
+	if (end) {
+		endDate = (end as any) instanceof Date ? (end as any) : parseISO(end);
+	}
```
Comment (nicolad) [inline assets/src/domain/eventEditor/ui/datetimes/dateForm/useDateFormConfig.ts]: looks like an utility function might be handy here

File: assets/src/application/ui/forms/espressoForm/utils.ts (hunk 1 above)
Comment (manzoorwanijk) [reply assets/src/application/ui/forms/espressoForm/utils.ts]: That's the reason I added the links in doc comment. You can see he shape of the objects from those links.

File: assets/src/domain/shared/services/utils/dateAndTime.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+};
+const transformTime = (value: any, originalValue: any): Date => {
+	return transformDateTime(value, originalValue, CONVERT_TO_MOMENT_TIME_FORMAT);
+};
+
+export const dateAndTimeSchema = yup.object({
+	startDate: yup
+		.date()
+		.transform(transformDate) // make sure we have Date object
+		.required(() => sprintf(__('%s is required'), __('Start Date'))),
```
Comment (tn3rb) [inline assets/src/domain/shared/services/utils/dateAndTime.ts]: as discussed in chat... let's just combine these strings and not use sprintf

File: assets/src/domain/shared/services/utils/dateAndTime.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		.transform(transformDate)
+		.required(() => sprintf(__('%s is required'), __('End Date')))
+		.when(['startDate'], (startDate: Date, schema: yup.DateSchema) => {
+			return schema.min(startDate, () => __('End Date & Time must be set later than the Start Date & Time'));
+		}),
+	endTime: yup
+		.date()
+		.transform(transformTime)
+		.required(() => sprintf(__('%s is required'), __('End Time')))
+		.when(
```
Comment (tn3rb) [inline assets/src/domain/shared/services/utils/dateAndTime.ts]: this looks really nice and clean 👍

File: assets/src/application/ui/forms/formModal/FormModalForm.tsx [hunk 6]
```diff
@@ -30,8 +30,10 @@ const FormModalForm: React.FC<FormModalFormProps> = ({
 	children,
 	title,
 }) => {
+	const submitDisabled = submitting || pristine || hasValidationErrors || hasSubmitErrors;
+
 	const submitButton: ButtonProps = {
-		disabled: submitting || pristine || hasValidationErrors || hasSubmitErrors,
+		disabled: submitDisabled,
```
Comment (nicolad) [inline assets/src/application/ui/forms/formModal/FormModalForm.tsx]: Nooooice :+1:

File: assets/src/domain/eventEditor/ui/datetimes/dateForm/useDateFormConfig.ts (hunk 3 above)
Comment (manzoorwanijk) [reply assets/src/domain/eventEditor/ui/datetimes/dateForm/useDateFormConfig.ts]: Done

File: assets/src/application/ui/forms/formModal/FormModalForm.tsx (hunk 2 above)
Comment (manzoorwanijk) [reply assets/src/application/ui/forms/formModal/FormModalForm.tsx]: Done

File: assets/src/application/ui/forms/espressoForm/utils.ts (hunk 1 above)
Comment (tn3rb) [reply assets/src/application/ui/forms/espressoForm/utils.ts]: the external links are great but what about a really simple basic before and after example as well?

File: assets/src/domain/shared/services/utils/dateAndTime.ts [hunk 7]
```diff
@@ -83,3 +83,10 @@ export const dateAndTimeSchema = yup.object({
 			}
 		),
 });
+
+export const prepareDateForForm = (mayBeDate: any, defaultDate: Date): Date => {
+	if (mayBeDate) {
+		return mayBeDate instanceof Date ? mayBeDate : parseISO(mayBeDate);
+	}
+	return defaultDate;
+};
```
Comment (nicolad) [inline assets/src/domain/shared/services/utils/dateAndTime.ts]: :+1:

File: assets/src/domain/eventEditor/ui/datetimes/dateForm/useDateFormConfig.ts [hunk 8]
```diff
@@ -19,16 +19,8 @@ const FIELD_NAMES: Array<keyof Datetime> = ['name', 'description', 'capacity', '
...
-
-	if (start) {
-		// "start" and "end" may be from Optimistic response as Date
-		startDate = (start as any) instanceof Date ? (start as any) : parseISO(start);
-	}
-	if (end) {
-		endDate = (end as any) instanceof Date ? (end as any) : parseISO(end);
-	}
+	const startDate = prepareDateForForm(start, PLUS_ONE_MONTH);
+	const endDate = prepareDateForForm(end, PLUS_TWO_MONTHS);
```
Comment (tn3rb) [inline assets/src/domain/eventEditor/ui/datetimes/dateForm/useDateFormConfig.ts]: 👍

File: assets/src/application/ui/forms/espressoForm/utils.ts [hunk 9]
```diff
@@ -83,6 +83,37 @@ export const evalFieldConditions = (conditions: FieldConditions, formData: AnyOb
...
+ * ]
+ *
+ * After error.inner.reduce, we will get:
+ * {
+ *     price: 'Price is required',
+ *     dateTime: {
+ *         startDate: 'Start date must not be in the past',
+ *         endDate: 'Start date must not be in the past',
+ *     },
+ * }
```
Comment (tn3rb) [inline assets/src/application/ui/forms/espressoForm/utils.ts]: great example 💯

**Pull request merged**